File: internal/service/balance.go
package service

import (
	"cmp"
	"slices"

	"danieleambrosino.it/opensettle/internal/types"
)

func ComputeBalances(obligations []types.Obligation) []types.Balance {
	balancesMap := make(map[types.PersonID]types.Cents)
	for _, o := range obligations {
		balancesMap[o.To] += types.Cents(o.Amount)
		balancesMap[o.From] -= types.Cents(o.Amount)
	}
	balances := make([]types.Balance, 0, len(balancesMap))
	for id, amt := range balancesMap {
		balances = append(balances, types.Balance{
			Person: id,
			Amount: amt,
		})
	}
	slices.SortFunc(balances, func(a, b types.Balance) int {
		comparison := cmp.Compare(a.Amount, b.Amount)
		if comparison != 0 {
			return comparison
		}
		return cmp.Compare(a.Person, b.Person)
	})
	return balances
}

File: webapp/src/lib/balance.ts
import type { Balance, Cents, Obligation, PersonID } from "./types";

export function computeBalances(obligations: Obligation[]): Balance[] {
  const balancesMap = new Map<PersonID, Cents>();
  for (const o of obligations) {
    balancesMap.set(o.to, (balancesMap.get(o.to) || 0) + o.amount);
    balancesMap.set(o.from, (balancesMap.get(o.from) || 0) - o.amount);
  }
  return balancesMap
    .entries()
    .map(([id, amt]) => ({ amount: amt, person: id }))
    .toArray()
    .toSorted(
      (a, b) => a.amount - b.amount || a.person.localeCompare(b.person)
    );
}
